### PR TITLE
Automaticaly format audio to ogg/opus

### PR DESCRIPTION
**Pull request by Helvio88**
# PR Details

Sending audio through wwebjs is not very reliable, as WhatsApp needs a very specific format

## Description

Added ffmpeg conversion in the Util class and automatic detection of mime for efficient conversion in sendMessage()

## Related Issue

https://github.com/pedroslopez/whatsapp-web.js/issues/1625
https://github.com/pedroslopez/whatsapp-web.js/issues/1584
https://github.com/pedroslopez/whatsapp-web.js/issues/1813

## Motivation and Context

I wanted to seamlessly send an audio via wwebjs like it happens with Stickers.
I am using ffmpeg to convert an mp3 file into ogg/opus and it works.
Wanted everyone to benefit from it, too.

## How Has This Been Tested

I have not tested the code I sent you, but I have done the conversion in a similar fashion in my code, essentially proxying all the Audio and PTT messages through it. It works

## Types of changes

- [ ] Dependency change
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [X] My code follows the code style of this project.
- [ ] I have updated the documentation accordingly (index.d.ts).

### CODE REVIEW

**Comment by Helvio88** [comment]
May I add, I couldn't find relevant documentation to update, only HTML that seemed to be autogenerated.

File: src/util/Injected.js [hunk 1]
```diff
@@ -13,7 +13,7 @@ exports.ExposeStore = (moduleRaidStr) => {
     window.Store.Cmd = window.mR.findModule('Cmd')[0].Cmd;
     window.Store.CryptoLib = window.mR.findModule('decryptE2EMedia')[0];
     window.Store.DownloadManager = window.mR.findModule('downloadManager')[0].downloadManager;
-    window.Store.Features = window.mR.findModule('FEATURE_CHANGE_EVENT')[0].LegacyPhoneFeatures;
+    window.Store.Features = window.mR.findModule('FEATURE_CHANGE_EVENT')[0]?.LegacyPhoneFeatures;
```
Comment (tuyuribr) [inline src/util/Injected.js:16]: Can't do this because node version

File: src/structures/GroupChat.js [hunk 2]
```diff
@@ -202,7 +202,7 @@ class GroupChat extends Chat {
             return window.Store.Invite.sendQueryGroupInviteCode(chatWid);
         }, this.id._serialized);
 
-        return code;
+        return code.code;
```
Comment (tuyuribr) [inline src/structures/GroupChat.js:205]: Probably from another pr

File: src/util/Injected.js [hunk 3]
```diff
@@ -552,17 +553,17 @@ exports.LoadUtils = () => {
...
-            await window.Store.ChatState.sendChatStateComposing(chatId);
-            break;
-        case 'recording':
-            await window.Store.ChatState.sendChatStateRecording(chatId);
-            break;
-        case 'stop':
-            await window.Store.ChatState.sendChatStatePaused(chatId);
-            break;
-        default:
-            throw 'Invalid chatstate';
```
Comment (tuyuribr) [inline src/util/Injected.js:565]: Please don't change things unrelated to this PR

File: src/util/Util.js [hunk 4]
```diff
@@ -174,6 +174,51 @@ class Util {
...
+     * Formats a media to ogg/opus
+     * @param {MessageMedia} media
+     * 
+     * @returns {Promise<MessageMedia> media} in opus/ogg (audio) format
+     */
+    static async formatToOpusAudio(media) {
+        if (!media.mimetype.includes('audio'))
+            throw new Error('media is not an audio');
+
+        if (media.mimetype === 'audio/ogg; codecs=opus')
```
Comment (tuyuribr) [inline src/util/Util.js:187]: Maybe a failsafe for all oggs codecs or waweb only like opus?

File: src/util/Util.js [hunk 5]
```diff
@@ -174,6 +174,51 @@ class Util {
...
+            return media
+
+        const tempFile = path.join(
+            tmpdir(), `${Crypto.randomBytes(6).readUIntLE(0, 6).toString(36)}.ogg`
```
Comment (tuyuribr) [inline src/util/Util.js:191]: Maybe use the same datadir from options (or do this in memory)?

File: src/util/Util.js [hunk 6]
```diff
@@ -174,6 +174,51 @@ class Util {
...
+            media.data.replace(`data:${media.mimetype};base64,`, ''),
+            'base64'
+        );
+        stream.push(buffer);
+        stream.push(null);
+
+        await new Promise((resolve, reject) => {
+            ffmpeg(stream)
+                .on('error', reject)
+                .on('end', () => resolve(true))
```
Comment (tuyuribr) [inline src/util/Util.js:205]: Maybe on error try send the bad format?

**Review by tuyuribr — CHANGES_REQUESTED** [review]

File: src/util/Util.js (hunk 5 above)
Comment (Helvio88) [reply src/util/Util.js:191]: ffmpeg does not work great in memory, and I followed the same format as other Util functions, so I think this should stay. Makes the whole class standardized and if someone wants to change all the temp locations, it's better to do it at once.

File: src/util/Injected.js (hunk 3 above)
Comment (Helvio88) [reply src/util/Injected.js:565]: This was probably to keep my code running, it'll be reversed in the next commit.

File: src/structures/GroupChat.js (hunk 2 above)
Comment (Helvio88) [reply src/structures/GroupChat.js:205]: This was probably to keep my code running, it'll be reversed in the next commit.

File: src/util/Injected.js (hunk 1 above)
Comment (Helvio88) [reply src/util/Injected.js:16]: This was probably to keep my code running, it'll be reversed in the next commit.

File: src/util/Util.js (hunk 4 above)
Comment (Helvio88) [reply src/util/Util.js:187]: This is the only WA accepted audio format if you want it to be played in the application, all other formats will be converted. Do you have a different sugestion? I'm keeping as is.

File: src/util/Util.js (hunk 6 above)
Comment (Helvio88) [reply src/util/Util.js:205]: Not possible. If you try to send the wrong format, while trying to send audio/ptt, it'll crash. The better solution is to throw an error and let this decision be made in a try-catch block, out of this routine.

**Pull request closed without merging**